client/http.go

```diff
@@ -0,0 +1,156 @@
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/drand/drand/beacon"
+	"github.com/drand/drand/key"
+	drand "github.com/drand/drand/protobuf/drand"
+)
+
+// HTTPClient is an interface for the exercised methods of an `http.Client`,
+// or equivlanet alternative.
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+	Get(url string) (resp *http.Response, err error)
+}
+
+// NewHTTPClient creates a new client pointing to an HTTP endpoint
+func NewHTTPClient(url string, groupHash []byte, client HTTPClient) (Client, error) {
+	c := &httpClient{
+		root:   url,
+		client: client,
+	}
+	group, err := c.FetchGroupInfo(groupHash)
+	if err != nil {
+		return nil, err
+	}
+	c.group = group
+
+	return c, nil
+}
+
+// NewHTTPClientWithGroup constructs an http client when the group parameters are already known.
+func NewHTTPClientWithGroup(url string, group *key.Group, client *http.Client) (Client, error) {
+	c := &httpClient{
+		root:   url,
+		group:  group,
+		client: client,
+	}
+	return c, nil
+}
+
+// httpClient implements Client through http requests to a Drand relay.
+type httpClient struct {
+	root   string
+	client HTTPClient
+	group  *key.Group
+}
+
+// FetchGroupInfo attempts to initialize an httpClient when
+// it does not know the full group paramters for a drand group.
+func (h *httpClient) FetchGroupInfo(groupHash []byte) (*key.Group, error) {
+	// fetch the `Group` to validate connectivity.
+	groupResp, err := h.client.Get(fmt.Sprintf("%s/group", h.root))
+	if err != nil {
+		return nil, err
+	}
+
+	protoGrp := drand.GroupPacket{}
+	if err := json.NewDecoder(groupResp.Body).Decode(&protoGrp); err != nil {
+		return nil, err
+	}
+	grp, err := key.GroupFromProto(&protoGrp)
+	if err != nil {
+		return nil, err
+	}
+
+	if grp.PublicKey == nil {
+		return nil, fmt.Errorf("Group does not have a valid key for validation")
+	}
+
+	if groupHash != nil && !bytes.Equal(grp.Hash(), groupHash) {
```

Comment:
Should it be an error if `groupHash` is `nil`?

Reply:
maybe worth an `insecure` flag? once can use a less secure client that only validates based on the TLS identity of the specified HTTP endpoint and starts with no other knowledge of the group identity

Reply:
I'll defer this to a level up: when making the http client as a functional option to the top level drand client, It seems useful to be able to specify something to the form of `WithInsecureHTTPEndpoint(url)` for testing and similar. In that case, a null grouphash can be provided. Otherwise `WithHTTPEndpoint(url)` will expect at least a valid groupHash is also already present when constructing the `HTTPClient`.